Enable SPAKE support where avilable

frozencemetery opened the pull request:
Because krb5 silently ignores unrecognized options, this is safe on
all versions.  It lands upstream in krb5-1.17; in Fedora, it was added
in krb5-1.6-17.

Upstream documentation can be found in-tree at
https://github.com/krb5/krb5/blob/master/doc/admin/spake.rst

frozencemetery commented:
CC @npmccallum, @simo5 (for awareness)

simo5 commented:
Very nice, but I wonder if we shouldn't also enable at last one of the NIST curves ?
Or are we confident all implementations will be able to suppor edwards25519 ?

frozencemetery commented:
edwards25519 is provided by krb5 itself, so it's always available.

simo5 commented:
I am talking about third party clients :-)

abbra commented:
We need to add some code to upgrade path. We already have augeas lense there in `ipaserver/install/server/upgrade.py` to perform upgrade to PKINIT, may be it can be extended.

tiran commented:
Thomas and I pointed out some issues with Augeas in #919. It doesn't handle kerberos config correctly. Is it possible to configure spake in e.g. ``/etc/krb5.conf.d/spake.conf`` instead of changing ``/etc/krb5.conf``?

```
# /etc/krb5.conf.d/spake.conf
[kdcdefaults]
  spake_preauth_kdc_challenge = edwards25519
```

abbra commented:
@tiran no, kdc.conf does not include /etc/krb5.conf.d.

frozencemetery commented:
@simo5 Per section 4: 

> Clients and KDCs MUST implement the edwards25519 group, but MAY choose not to offer or accept it by   default.

simo5 commented:
Ah thanks, I missed that it was a MUST.

simo5 commented:
@abbra actually afaik the KDC merges krb5.conf and kdc.conf in a single profile, so I surely hope it read /etc/krb5.conf.d

npmccallum commented:
Have any other krb5 implementations discussed implementing this?

frozencemetery commented:
@npmccallum Not to my knowledge, though we probably wouldn't know until they'd done it.

abbra commented:
@simo5  checked kdc code and yes, it uses profile loading code with a list of files to load, by adding the base krb5 configuration files to the base kdc configuration files. So yes, resulting input to a profile loader is something like /etc/krb5.conf:/var/kerberos/krb5kdc/kdc.conf  which means if krb5.conf includes /etc/krb5.conf.d/ files, they will be visible to kdc.conf too.
This should simplify the configuration also for PKINIT code.

frozencemetery commented:
Configuring krb5.conf.d is possible, but there's no code to do that in the freeipa client right now.  I've added some, but I'm not very familiar with this part of the code.  In general, there's a lot of stuff in the client that we should move out to krb5.conf.d and keep the amount of changes we make to krb5.conf proper small.

I've added upgrade logic as well.  Should I be checking for the krb5.conf.d file there as well, or will that be handled elsewhere?

tiran commented:
Could we ship the config files with freeipa-client package and have them installed on ``dnf install freeipa-client`` time? Or are the settings too intrusive?

frozencemetery commented:
@tiran That would probably be fine, but some of them will be duplicates of what I have in /etc/krb5.conf already (in the libkrb5 package).  Not an issue as long as we expect the number of people with freeipa-client installed who haven't run `ipa-client-install` to be small.

My concern is for the other downstreams - we'd want it to be clear that they need to package this file, and I don't know how reasonable that is.

simo5 commented:
I do not think changing krb5 behavior just because a package is installed is sensible, besides we need domain specific data, so we'd still need to add ad hoc config on top of static files. Better to use a tamplate but store the results in /etc/krb5.d

npmccallum commented:
What is the reason that we are trying to enable this specifically in FreeIPA? Can't we just let FreeIPA depend upon the OS's configuration? IMHO, that would be preferred for now since we don't require it for any specific FreeIPA feature.

frozencemetery commented:
For the server, there's no notion of OS configuration - we ship a default file, but it has to be customized for the realm.

For the client, I would love it if freeIPA would leave the OS configuration alone.  But right now it doesn't do that - anything I configure in krb5.conf as part of libkrb5 gets stomped on by `ipa-client-install`.

frozencemetery commented:
Updated.  If there are further changes you'd like to see, please let me know.

abbra commented:
LGTM. Thanks!

tiran commented on ipaclient/install/client.py:
```diff
@@ -627,6 +627,16 @@ def configure_krb5_conf(
+    template = os.path.join(
+        paths.USR_SHARE_IPA_DIR,
+        os.path.basename(paths.KRB5_FREEIPA) + ".template")
+    contents = open(template).read()
```
Use ``shutil.copyfile()`` to copy files.

npmccallum commented:
Congrats everyone! This has been a long time coming. Thanks to everyone who has been a part of this process!

tiran commented:
master:

* 792adebfabb456d154164387fb7e60acb30f4325 Enable SPAKE support using krb5.conf.d snippet
* 73c3495db2dc1df38407aa7b49fed2f222e85bbe Use shutil to copy file
